### PR TITLE
ca: remove one call to Provider.ActiveRoot

### PR DESCRIPTION
**Pull request by dnephin**
This change is related to #11598, and #11159.  More detail in the commit messages, and I'll call out a few things below with inline comments.

The PR is not large, but it is still best viewed by individual commit, as there are a few independent changes, and the commit message has more context.

ActiveRoot should not be called from the secondary DC, because there should not be a requirement to run the same Vault instance in a secondary DC (#11159). `SignCertificate` is called in a secondary DC, so it should not call `Provider.ActiveRoot`. This PR removes that call.

We would also like to change the interface of `Provider.ActiveRoot` so that we can support using an intermediate cert as the primary CA in Consul (#11598). In preparation for making that change I am reducing the number of calls to `Provider.ActiveRoot`, so that there are fewer code paths to modify when the interface changes.

In order to make this change I had to modify the existing test a bit, and fix a couple problems with the mock. I also reduced the runtime of the test so that I wasted less time waiting for it to run.

### CODE REVIEW

File: agent/structs/connect_ca.go
Comment (dnephin) [inline agent/structs/connect_ca.go:102]: This convention of "leaf signing cert last" already existed for the new cert expiry metrics we added. I'm calling it out here more explicitly so that it is well understood. We are going to be depending on this convention even more as part of subsequent changes for #11598.

File: agent/consul/leader_connect_ca.go
Comment (dnephin) [inline agent/consul/leader_connect_ca.go:1506]: Calling out this change loudly. I'm pretty sure this is safe, but I would appreciate some close attention to be sure I'm not missing something. Echo'ing the commit message here:

> Immediately above this line we are already appending the full list of
intermediates. The `provider.ActiveIntermediate` MUST be in this list of
intermediates because it must be available to all the other non-leader
Servers.  If it was not in this list of intermediates then any proxy
that received data from a non-leader would have the wrong certs.

> This is being removed now because we are planning on changing the
`Provider.ActiveIntermediate` interface, and removing these extra calls ahead of
time helps make that change easier.

From what I could tell in git history this logic was part of the initial implementation, so likely just an oversight that we haven't noticed yet.

File: agent/consul/leader_connect_ca_test.go
Comment (dnephin) [inline agent/consul/leader_connect_ca_test.go:244]: This is definitely weird, but it preserves the old behaviour of this test. I think we'll need to do a bit more work to make this mock a little more realistic.

File: agent/consul/leader_connect_ca_test.go
Comment (dnephin) [inline agent/consul/leader_connect_ca_test.go:403]: This was just to make debugging easier. It should not change any functionality.

File: agent/consul/leader_connect_ca.go
Comment (dnephin) [reply agent/consul/leader_connect_ca.go:1506]: Hmm, because of the bug described in #10532, this change is not safe. We'd have to fix that first for this change to be ok.

Comment (dnephin) [reply agent/consul/leader_connect_ca.go:1506]: #11671 fixes the problem that this is masking, so rebasing this PR on that change makes this safe once again.

File: agent/consul/leader_connect_ca_test.go
Comment (kisunji) [inline agent/consul/leader_connect_ca_test.go:439]: Confirming this change is intentional?

**Review by kisunji — APPROVED** [review]
one minor question, but LGTM (I assume you can't merge until the linked issue is resolved)

File: agent/consul/leader_connect_ca.go
Comment (FFMMM) [inline agent/consul/leader_connect_ca.go:1547]: so if we are not the primary DC we always sign with the intermediate?

File: agent/consul/leader_connect_ca_test.go
Comment (dnephin) [reply agent/consul/leader_connect_ca_test.go:439]: Oh ya, great question! I should have given more of an explanation of this part of the change.

I was running into a panic in this test because of this line:
https://github.com/hashicorp/consul/blob/1a7b89c6cbb82ab497963846ce7e8fa39ca3b53a/agent/consul/leader_connect_ca.go#L1532

Basically the PEM encoded string coming out did not have any URIs. I tracked it back to how our fake was using the wrong certs for the local intermediate cert. Previously we would return the root cert as the "intermediate cert", and the root cert happened to have URIs on it, so the test was passing.

So once I fixed the fake to return the correct cert, I had to fix the generation of that cert to include the URIs.

The fake is still a bit strange. Normally the signed leaf cert should be the one with URIs on it, but the fake is returning an empty string for the actual leaf cert, so we have to rely on the URIs in one of the parents.

I think we need to do more to clean this up, but I didn't want to include too much test cleanup in this PR.

File: agent/consul/leader_connect_ca.go
Comment (dnephin) [reply agent/consul/leader_connect_ca.go:1547]: Yup, exactly! Every provider uses an intermediate certificate in a secondary DC to sign leaf certs.

File: agent/consul/leader_connect_ca.go
Comment (dnephin) [inline agent/consul/leader_connect_ca.go:1475]: Leaving a TODO for myself. This can use `getLeafSigningCertFromRoot`  once #11671 is merged.

```suggestion
	if c.isIntermediateUsedToSignLeaf() && len(caRoot.IntermediateCerts) > 0 {
		inter := c.getLeafSigningCertFromRoot(caRoot)
```

**Review by kyhavlov — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/553249.

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit f31e0b8b1a6bd5554c684a5781324202e9143d86 onto `release/1.11.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/568392)

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit f31e0b8b1a6bd5554c684a5781324202e9143d86 onto `release/1.10.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/568392)

**Pull request merged**
